includes/class-wp-job-manager-post-types.php

```diff
@@ -691,49 +692,83 @@ public function get_filled_job_listings(): array {
 	}
 
 	/**
-	 * Get expired jobs.
+	 * Maybe exclude filled job listings from search and archive pages.
 	 *
-	 * @return array
+	 * @param $query WP_Query $query
+	 *
+	 * @return void
 	 */
-	public function get_expired_job_listings(): array {
-		if ( ! get_option( 'job_manager_hide_expired' ) ) {
-			return [];
+	public function maybe_hide_filled_job_listings( WP_Query $query ): void {
+		$hide_filled_positions = get_option( 'job_manager_hide_filled_positions' );
+
+		if ( ! $hide_filled_positions ) {
+			return;
 		}
 
-		$expired_jobs_transient = get_transient( 'hide_expired_jobs_transient' );
-		if ( false === $expired_jobs_transient ) {
-			$expired_jobs_transient = get_posts(
-				[
-					'post_type'   => 'job_listing',
-					'post_status' => 'expired',
-					'fields'      => 'ids',
-				]
-			);
-			set_transient( 'hide_expired_jobs_transient', $expired_jobs_transient, DAY_IN_SECONDS );
+		if (
+			! is_admin()
+			&& $query->is_main_query()
+			&& ( $query->is_search() || $query->is_archive() )
+		) {
+
+			$query->set( 'post__not_in', $this->get_filled_job_listings() );
 		}
-		return $expired_jobs_transient;
 	}
 
 	/**
-	 * Maybe exclude expired and/or filled job listings from search.
+	 * Maybe exclude expired job listings from search and archive pages.
 	 *
 	 * @param $query WP_Query $query
 	 *
 	 * @return void
 	 */
-	public function maybe_hide_filled_expired_job_listings_from_search( WP_Query $query ): void {
-		$hide_filled_positions = get_option( 'job_manager_hide_filled_positions' );
-		$hide_expired          = get_option( 'job_manager_hide_expired' );
+	public function maybe_hide_expired_job_listings( WP_Query $query ): void {
+		$hide_expired = get_option( 'job_manager_hide_expired' );
 
-		if ( ! $hide_filled_positions && ! $hide_expired ) {
+		if ( ! $hide_expired ) {
 			return;
 		}
 
-		if ( ! is_admin() && $query->is_search() ) {
-			$jobs_to_exclude = array_merge( $this->get_filled_job_listings(), $this->get_expired_job_listings() );
-			if ( ! empty( $jobs_to_exclude ) ) {
-				$query->set( 'post__not_in', $jobs_to_exclude );
-			}
+		if (
+			! is_admin()
+			&& $query->is_main_query()
+			&& ( $query->is_search() || $query->is_archive() )
+		) {
```

Comment:
Same here about the comment explaining the logic.

Reply:
Committed https://github.com/Automattic/WP-Job-Manager/pull/2449/commits/27b3be7732a15d5d95d1981dbbb6f7ddceb4a009